### DISCUSSION TITLE
Proposal: move Pex out of the Pantsbuild org.

### DISCUSSION THREAD

**Discussion by jsirois** [pex-tool/pex#2354] [General]
This is driven by me being the ~sole maintainer and having a harder and harder time dealing with Pantsbuild contributions and interactions charitably. Pantsbuild should not have to suffer my quirks in particular and Pex has no real tie to Pants or Pantsbuild except via history.

Concretely, I'd like to:
+ Move away from Pantsbuild slack (I use Linen to poll for Pex support issues currently).
+ Define Pex goals, contribution guidelines, CoC, etc. crisply and cleanly.

I can see two paths forward that meet these goals:
1. Transfer pex to a new github org.
2. Fork pex to a new github org and continue my development / maintainance / releases (including to PyPI) there.

Either of these options requires Pantsbuild org buy in; so I'm looking for your feedback.

I'll seed this with a few mentions of folks who can pull in others if they wish @benjyw, @stuhood, @kaos, @huonw, @tgolsson and @thejcannon.

**Comment by tgolsson** [comment]
I don't know enough about the history of Pants and Pex to know why there's friction here. To me it seems like if you don't want to develop Pex under the Pantsbuild banner, noone can stop you from doing it elsewhere. I haven't gotten the impression that Pantsbuild as an organization gets very involved in the development of pex, or puts lots of requirements on how it's ran. 

Either way, pex seems to me like such an integral part of pants at this point that it cannot be easily replaced in the short term. I realize this dependence probably adds to your work, and being part of the organization puts you in the blast zone via association. Still, even if we part ways Pantsbuild is going to expose a lot of people to pex, and issues will be filed in relation to that. Maybe there'll be less pressure on you to solve them, fewer pings, and maybe we'll be more inclined to solve them from the pants end. However, I've also felt that pex is yours and onboarding others isn't in your interest, nor that I could contribute accordingly.

So no matter the "physical" location of the project, I'm more interested in what we can do to cooperate better, if that's the issue. Maybe moving pex is enough, but if there's other concerns I think those are worth addressing separately, here or elsewhere. But as said, I'm still "new" (and thus maybe a demanding user ;-))

**Reply by jsirois** [reply]
> I don't know enough about the history of Pants and Pex to know why there's friction here.

I don't like the Pants community as it has evolved. I wanted to make a good build tool and that's about all. I was interested in working with people who wanted the same. Benjy joined in early in 2011 in that quest and Stu a year or two later. Since that time a broader community has grown and I think almost everyone agrees this has been for the better. I am probably the lone person out seeing negatives here, but I do see them, and my natural inclination to point them out just leads to pain. I have no wish to cause the undue pain I currently cause in the community; so best to step away. Pex being tied into the org / CoC / Slack, etc is the last thing preventing me from stepping away and not imposing this pain.


> To me it seems like if you don't want to develop Pex under the Pantsbuild banner, noone can stop you from doing it elsewhere.

That is basically true; however, the PyPI project has a single logical owner, which is Pantsbuild. For this reason, at the very least, some agreed decision needs to be made of who / where the `pex` PyPI project can be canonically published from.

> So no matter the "physical" location of the project, I'm more interested in what we can do to cooperate better, if that's the issue.

That's not the issue. I get great bug reports, diagnosis, etc from you, @huon, @cburroughs, @benjyw etc.

**Reply by jsirois** [reply]
@tgolsson maybe this bit of perspective helps: When I came up with the idea of scie ~16 months ago I discussed this with other Toolchain folks. Of the 3 places to house the scie projects we discussed, Pantsbuild, new org, or Toolchain, I only pushed back against Pantsbuild. As such, we agreed on a new org. I'm not sure my reasons for resistance to Pantsbuild as a home were discussed at that time, but it was driven by these same considerations, probably less strongly felt, less clear in my mind, at the time.

**Reply by tgolsson** [reply]
Thanks for all the clarifications. My only concern then is bus-factor/break-glass type situations if you're the (sole) owner of that organization -- maybe something you've thought about or there's arrangements in place from scie, but feel like I'd bring it up.

(And I'm fully in-favor, either way.)

**Reply by jsirois** [reply]
That is a valid concern, but I think its covered by the current situation on PyPI though (see the maintainers list): https://pypi.org/project/pex/

The repo can be forked for example and the fork become canonical in a bus situation; so really the only critical issue is that there are multiple publishing keys for PyPI, which there are.

**Comment by benjyw** [comment]
I'm generally of the opinion that whoever puts in the legwork gets to make the decisions. This is partly out of courtesy, but mostly because the person doing the work has the best perspective on those decisions. So as the ~sole maintainer, if this makes sense to you, I support it. 

I would greatly prefer transferring the project to another org to forking. Pants already consumes Pex as a third-party binary, and Pex is highly scrupulous about maintaining backwards compatibility, so this doesn't impose much hardship on Pants that I can see. As you mention, we already work this way with scie, so I hardly think this is an outlandish state to be in... 

I believe that Pants is the major (although not exclusive) consumer of Pex? So most bug reports, feature requests, contributions, code reviews and so on will, I would guess, tend to still involve people in the Pants community. Those folks (myself included) will now have to operate under a different set of guidelines. Which is totally fine, and we'll all get used to it pretty quickly, but in the short term there will probably be some confusion, misunderstandings or friction. So let's make sure we resolve those with as little tension as possible until we settle in to the new normal.

**Reply by sureshjoshi** [reply]
I'm also of a legwork-driven-decision-making structure. 

If the rest of the Pants-team were regular code contributors/maintainers, that alternate reality could lead to a different discussion - but as John's the (effectively) sole maintainer, I'd also defer to his opinion/preference on how best to structure, enhance, and maintain Pex. 

My two cents on the paperwork side of that discussion would be (1) transferring to a new org, rather than the fork and maintain strat.

**Reply by kaos** [reply]
I agree with the above.

**Reply by jsirois** [reply]
Thank you all for the feedback so far.

I also favor option 1 (transfer to a new org). I'll proceed on Friday 2/9 unless sentiment diverges in the intervening days.

**Comment by jsirois** [comment]
Ok, this is done and the repo is now housed at https://github.com/pex-tool/pex. There will be some follow-up activity over the next few days to cleanup docs, etc. Thanks everyone.

**Comment by jsirois** [comment]
@hrfuller, @Eric-Arellano and @stuhood: Pex has moved here to https://github.com/pex-tool/pex. I'm beginning a clean up process and part of that is auditing PyPI access / making sure publishing still works etc. You three are maintainers on PyPI, but not active maintainers in reality. My inclination is to trim you all off that list. I'll proceed to do so later next week unless I hear objections.

**Reply by hrfuller** [reply]
That's fine with me, thanks John.

**Reply by Eric-Arellano** [reply]
That's also fine with me. Thanks!

**Reply by jsirois** [reply]
Ok. This is now complete.

**Comment by jsirois** [comment]
Ok, and as part of my 1st pass audit, pantsbuild/pants has no references to pantsbuild/pip - this is just used for the home of Pex's vendored / patched Pip 20.3.4. I'll transfer that repo to the pex-tool org as well.

**Comment by jsirois** [comment]
And https://github.com/pantsbuild/setuptools is likewise just used by Pex for a vendored setuptools patch. I'll move that as well.

**Comment by jsirois** [comment]
Thanks folks. As of v2.2.1 I consider the project fully migrated and a bit improved with the new doc site and embedded docs.
